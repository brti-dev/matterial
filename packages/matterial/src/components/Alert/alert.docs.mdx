---
name: Alert
description: Error, warning, and other important notes
---

## Usage

```jsx
import { Alert } from 'matterial'

function AlertExample() {
  return <Alert severity="error" message="Something happened :(" />
}
```

## Examples

<Alert severity="error" label="Critical Error" type="assertive">
  Something happened
</Alert>
<Alert
  severity="warning"
  message="Danger!"
  action={
    <Button variant="contained" color="warning">
      Take me to safety
    </Button>
  }
/>
<Alert severity="success" label="OK" message="Well done" />
<Alert severity="info" action={<Link href="/">Do Lorem</Link>}>
  Velit laborum quis excepteur laboris commodo fugiat incididunt eiusmod cillum
  culpa sit id ipsum ex.
</Alert>
<Alert dismiss>
  Velit do velit est cillum reprehenderit cillum aliqua quis officia velit
  deserunt.
</Alert>

## Colors

{SEVERITIES.map(severity => <Alert severity={severity} label={false}>{severity}</Alert>)}

# `useAlert` Hook

A React Hook to access an alert state and component.

## Usage

```jsx
function App() {
  const [AlertComponent, setAlert] = useAlert()

  return (
    <AlertComponent />
    <button onClick={() => setAlert('Something happened')}>Alert</button>
    <button onClick={() => setAlert({ message:'Danger!', severity: 'warning' })}>Danger</button>
    <button onClick={() => setAlert(null)}>Reset</button>
  )
}
```
