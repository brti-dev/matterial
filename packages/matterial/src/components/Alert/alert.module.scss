@use '../../styles/breakpoint' as *;

.alert {
  display: flex;
  align-items: flex-start;
  gap: 0.5em;
  padding: 0.5em;
  border: var(--border-width) solid var(--color-border);
  border-radius: var(--radius);
  color: var(--color-foreground);
  line-height: 1.43;
  letter-spacing: 0.01071em;
  background: var(--color-background);

  .icon {
    display: flex;

    & > * {
      width: 1.2rem;
      height: 1.2rem;
    }
  }

  .label {
    font-weight: 800;
    text-transform: uppercase;
  }

  .content {
    width: 100%;
    display: flex;
    gap: 0.5em;
    justify-content: space-between;
    align-items: baseline;
    font-size: 0.85rem;
    font-weight: 400;

    .message {
      flex-basis: 100%;
      word-break: break-word;
    }

    .action {
      flex-shrink: 0;
    }

    @include breakpoint('mobile') {
      flex-direction: column;
    }
  }

  & + .alert {
    margin-top: 1em;
  }
}
