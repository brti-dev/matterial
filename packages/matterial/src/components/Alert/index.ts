export * from './alert'
export * from './alert.example'
export * from './use-alert'
