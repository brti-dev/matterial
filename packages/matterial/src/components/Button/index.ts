export * from './button'
export * from './icon-button'
export { default } from './button'
