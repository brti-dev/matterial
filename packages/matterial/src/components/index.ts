export { default as Divider } from './Divider'
export { default as DateTime } from './DateTime'
