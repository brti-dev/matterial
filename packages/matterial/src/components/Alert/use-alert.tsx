import { useCallback, useReducer } from 'react'

import Alert, { AlertDispatch } from 'components/Alert'

export function reducer(
  _: /* state */ null | AlertDispatch,
  action?: null | string | AlertDispatch
): null | AlertDispatch {
  if (!action) {
    return { message: null }
  }

  if (typeof action === 'string') {
    return { message: action }
  }

  return action
}

/**
 * Hook to manage alert state.
 *
 * @param initialState Initial state; If true-ish, the alert will display immediately
 *
 * @returns {JSX.Element} Alert component to render
 * @returns {function} Function to set alert message
 */
export function useAlert(
  initialState?: string | AlertDispatch
): [React.ComponentType, any] {
  const [alert, setAlert] = useReducer(reducer, reducer(null, initialState))

  const component = useCallback(() => {
    if (!alert?.message) return null

    return <Alert {...{ ...alert }} />
  }, [alert])

  return [component, setAlert]
}
