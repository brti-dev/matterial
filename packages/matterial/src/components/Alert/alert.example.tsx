import { useAlert } from '.'
import Button from 'components/Button'

export function AlertExample() {
  const [AlertComponent, setAlert] = useAlert()

  return (
    <>
      <AlertComponent />
      <div style={{ display: 'flex', gap: '1em', margin: '1em 0' }}>
        <Button
          variant="outlined"
          color="error"
          onClick={() => setAlert('Something happened')}
        >
          Alert
        </Button>
        <Button
          variant="outlined"
          color="warning"
          onClick={() => setAlert({ message: 'Danger!', severity: 'warning' })}
        >
          Danger
        </Button>
        <Button variant="outlined" onClick={() => setAlert(null)}>
          Dismiss
        </Button>
      </div>
    </>
  )
}
