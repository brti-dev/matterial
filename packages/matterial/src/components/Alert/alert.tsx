import * as React from 'react'
import ReachAlert from '@reach/alert'
import { ErrorIcon, WarningIcon, SuccessIcon, InfoIcon } from '../Icons'

import { Severity, Variant } from '../../interfaces/theme'
import classnames from '../../lib/classnames'
import cssColor from '../../lib/css-color'
import { Button } from '../Button'
import classes from './alert.module.scss'

const LABELS = {
  error: 'Error',
  warning: 'Warning',
  success: 'Success',
}
const ICONS = {
  error: <ErrorIcon />,
  warning: <WarningIcon />,
  success: <SuccessIcon />,
  info: <InfoIcon />,
}

type Icon = boolean | React.ReactElement

export type AlertDispatch = {
  /**
   * A button or other call to action
   */
  action?: string | React.ReactElement
  /**
   * Append an action to dismiss the alert; Overwritten by `action`
   */
  dismiss?: boolean
  /**
   * Display an icon; Differs by severity, or indicate custom; Default: false
   */
  icon?: Icon
  /**
   * Prefix a short phrase; Defaults to a phrase based on severity, or suppress with false
   */
  label?: boolean | string
  /**
   * A short message to show; Nullable for use in useAlert hook
   */
  message: string | null
  /**
   * Describes the type of alert
   */
  severity?: Severity
  /**
   * Controls whether the assistive technology should read immediately
   * ("assertive") or wait until the user is idle ("polite"); Default: 'polite'
   */
  type?: 'polite' | 'assertive'
  /**
   * Style variants
   */
  variant?: Variant
}

export type AlertProps = Partial<AlertDispatch> & {
  children?: React.ReactNode
  className?: string
  style?: React.CSSProperties
}

/**
 * Regenerate a React element with the prop `size` small
 */
function shrink(component: string | React.ReactElement) {
  if (React.isValidElement(component)) {
    return React.cloneElement(component, { size: 'small' })
  }

  return component
}

export function Alert({
  action,
  children,
  className,
  dismiss = false,
  icon = false,
  label: naturalLabel = true,
  message: naturalMessage,
  severity,
  style: naturalStyle = {},
  type,
  variant = 'outlined',
}: AlertProps): JSX.Element {
  const classNames = classnames(
    classes.alert,
    `variant--${variant}`,
    severity && `color--${severity}`,
    className
  )

  let [message, setMessage] = React.useState(children || naturalMessage)

  if (dismiss && !action) {
    action = (
      <Button
        variant="outlined"
        color={severity}
        onClick={() => setMessage(null)}
      >
        Dismiss
      </Button>
    )
  }

  if (!message) {
    return <></>
  }

  let label: any
  if (naturalLabel === true && severity) {
    // @ts-ignore
    label = severity in LABELS ? LABELS[severity] : null
  } else if (typeof naturalLabel === 'string') {
    label = naturalLabel
  }

  const color = severity || 'default'
  const style = { ...naturalStyle, '--color': cssColor(color) }

  return (
    <ReachAlert
      type={type}
      className={classNames}
      role="alert"
      // aria-label={label || severity || 'alert'}
      data-severity={severity}
      style={style}
    >
      <>
        {icon && <Icon icon={icon} severity={severity} />}
        <div className={classes.content}>
          <div className={classes.message}>
            {!!label && <strong className={classes.label}>{label}: </strong>}
            {message}
          </div>
          {action && <div className={classes.action}>{shrink(action)}</div>}
        </div>
      </>
    </ReachAlert>
  )
}

function Icon({
  icon,
  severity,
}: {
  icon: Icon
  severity?: Severity
}): JSX.Element | null {
  if (icon === true) {
    // Return icon by severity, or default icon
    if (!severity || !ICONS[severity]) {
      severity = 'info'
    }
    return (
      <div className={classes.icon} aria-hidden="true">
        {ICONS[severity]}
      </div>
    )
  }

  // Return custom icon
  return (
    <div className={classes.icon} aria-hidden="true">
      {icon}
    </div>
  )
}
