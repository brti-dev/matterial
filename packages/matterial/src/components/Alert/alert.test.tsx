import '@testing-library/jest-dom'
import renderer from 'react-test-renderer'
import userEvent from '@testing-library/user-event/dist'

import { Severity, Variant } from 'interfaces/theme'
import { reducer } from 'components/Alert/use-alert'
import { render, screen } from '../../../test-utils'
import { Alert, AlertDispatch } from 'components/Alert'
import { Button } from 'components/Button'
import { AlertExample } from '.'

const SEVERITY = ['error', 'warning', 'info', 'success']

describe('useAlert hook', () => {
  const initialState: AlertDispatch = {
    label: 'foo',
    message: 'lorem',
    severity: 'info',
  }

  test('should set up reducer', () => {
    expect(reducer(null, initialState)).toEqual(initialState)
  })

  test('alert appears and disappears', () => {
    render(<AlertExample />)

    const alertBtn = screen.getByText('Alert')
    const dismissBtn = screen.getByText('Dismiss')

    expect(alertBtn).toBeInTheDocument()
    expect(dismissBtn).toBeInTheDocument()
    expect(screen.queryByRole('alert')).not.toBeInTheDocument() // Query will return node OR null; It will never throw

    userEvent.click(alertBtn)
    expect(screen.queryByRole('alert')).toBeInTheDocument()

    userEvent.click(dismissBtn)
    expect(screen.queryByRole('alert')).not.toBeInTheDocument()
  })
})

test('should render correctly', () => {
  const tree = renderer
    .create(
      <Alert severity="error" action={<b>Action</b>}>
        Alert
      </Alert>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('should work with different variants', () => {
  const tree = renderer
    .create(
      <>
        {['default', 'outlined', 'contained'].map((variant: Variant) => (
          <Alert variant={variant} key={variant}>
            {variant}
          </Alert>
        ))}
      </>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('has the proper `severity` attributes', () => {
  const severity = SEVERITY[0]
  const { getByRole, rerender } = render(
    <Alert severity={severity as Severity} label="Alert">
      Foo
    </Alert>
  )
  expect(getByRole('alert')).toHaveAttribute('data-severity', severity)

  SEVERITY.slice(1).forEach((severity: Severity) => {
    rerender(<Alert severity={severity}>alert</Alert>)
    expect(getByRole('alert')).toHaveAttribute('data-severity', severity)
  })
})

test('should output a message using the `message` prop', () => {
  const { getByText } = render(<Alert message="WARNING!" />)
  expect(getByText('WARNING!')).toBeInTheDocument()
})

test('should output a message using the `children` prop', () => {
  const { getByText } = render(<Alert>WARNING!</Alert>)
  expect(getByText('WARNING!')).toBeInTheDocument()
})

test('should include a call to action with the `action` prop', () => {
  const { getByRole } = render(
    <Alert message="Alert" action={<Button>Act</Button>} />
  )
  expect(getByRole('button')).toBeInTheDocument()
})

test('should dismiss', () => {
  const { getByRole } = render(<Alert dismiss>Alert</Alert>)

  const btn = getByRole('button')
  expect(btn).toBeInTheDocument()
  userEvent.click(btn)
  expect(btn).not.toBeVisible()
})
