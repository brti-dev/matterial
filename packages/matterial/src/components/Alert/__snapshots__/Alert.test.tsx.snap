// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="alert variant--outlined color--error"
  data-reach-alert={true}
  data-severity="error"
  role="alert"
>
  <div
    aria-hidden="true"
    className="icon"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M11.953,2C6.465,2,2,6.486,2,12s4.486,10,10,10s10-4.486,10-10S17.493,2,11.953,2z M12,20c-4.411,0-8-3.589-8-8 s3.567-8,7.953-8C16.391,4,20,7.589,20,12S16.411,20,12,20z"
      />
      <path
        d="M11 7H13V14H11zM11 15H13V17H11z"
      />
    </svg>
  </div>
  <div
    className="content"
  >
    <div
      className="message"
    >
      Alert
    </div>
    <div
      className="action"
    >
      <b
        size="small"
      >
        Action
      </b>
    </div>
  </div>
</div>
`;

exports[`should work with different variants 1`] = `
Array [
  <div
    className="alert variant--default"
    data-reach-alert={true}
    role="alert"
  >
    <div
      className="content"
    >
      <div
        className="message"
      >
        default
      </div>
    </div>
  </div>,
  <div
    className="alert variant--outlined"
    data-reach-alert={true}
    role="alert"
  >
    <div
      className="content"
    >
      <div
        className="message"
      >
        outlined
      </div>
    </div>
  </div>,
  <div
    className="alert variant--contained"
    data-reach-alert={true}
    role="alert"
  >
    <div
      className="content"
    >
      <div
        className="message"
      >
        contained
      </div>
    </div>
  </div>,
]
`;
